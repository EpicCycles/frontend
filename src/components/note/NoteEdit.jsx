import React from "react";
import {updateObject} from "../../helpers/utils";
import EditModelPage from "../app/model/EditModelPage";
import {customerNoteFields} from "../app/model/helpers/fields";
import {addFieldToState, getModelKey} from "../app/model/helpers/model";
import ModelEditIcons from "../app/model/ModelEditIcons";

class NoteEdit extends React.Component {
    state = {};

    componentWillMount() {
        this.setState({ note: updateObject(this.props.note) });
    };


    handleInputChange = (fieldName, input) => {
        const updatedNote = addFieldToState(this.state.note, customerNoteFields, fieldName, input);
        this.setState({ note: updatedNote });
    };

    onClickReset = () => {
        if (this.state.note.id) {
            this.setState({ note: updateObject(this.props.note) });
        } else {
            this.setState({ note: {}});
        }
    };

    render() {
        const { note } = this.state;
        const { saveNote, deleteNote } = this.props;

        return <div>
            <EditModelPage
                modelFields={customerNoteFields}
                onChange={this.handleInputChange}
                model={note}
                persistedModel={this.props.note}
            />
            <div className="align_right">
                <ModelEditIcons
                    componentKey={getModelKey(note)}
                    model={note}
                    modelDelete={deleteNote}
                    modelReset={this.onClickReset}
                    modelSave={saveNote}
                />
            </div>
        </div>
    }
}

NoteEdit.defaultProps = {
    note: {},
};
export default NoteEdit;
