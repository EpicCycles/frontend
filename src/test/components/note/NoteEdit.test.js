import React from 'react';
import toJson from 'enzyme-to-json';
import NoteEdit from '../../../components/note/NoteEdit';

describe('NoteEdit tests', () => {
  const note = {
    id: 23,
    note_text: 'note text here',
    customer_visible: true,
  };
  const noteNoId = {
    note_text: 'note text here',
    customer_visible: true,
    changed: true,
  };
  it('renders the form text correctly with no note', () => {
    const saveNote = jest.fn();
    const deleteNote = jest.fn();
    const removeNote = jest.fn();

    const input = shallow(
      <NoteEdit saveNote={saveNote} deleteNote={deleteNote} removeNote={removeNote} />,
    );
    expect(toJson(input)).toMatchSnapshot();
  });
  it('renders the form text correctly with note', () => {
    const saveNote = jest.fn();
    const deleteNote = jest.fn();
    const removeNote = jest.fn();

    const input = shallow(
      <NoteEdit saveNote={saveNote} note={note} deleteNote={deleteNote} removeNote={removeNote} />,
    );
    expect(toJson(input)).toMatchSnapshot();
  });
  it('clears data when no note and reset is clicked', () => {
    const saveNote = jest.fn();
    const deleteNote = jest.fn();
    const removeNote = jest.fn();
    const input = shallow(
      <NoteEdit
        saveNote={saveNote}
        note={noteNoId}
        deleteNote={deleteNote}
        removeNote={removeNote}
      />,
    );
    expect(input.state('note')).toEqual(noteNoId);

    input.instance().handleInputChange('note_text', 'big note text');
    input.instance().handleInputChange('customer_visible', true);
    expect(input.state('note')).not.toEqual(noteNoId);

    input.instance().onClickReset();
    expect(input.state('note')).toEqual({});
  });
  it('resets to passed data when note and reset is clicked', () => {
    const saveNote = jest.fn();

    const input = shallow(<NoteEdit note={note} saveNote={saveNote} />);
    expect(input.state('note')).toEqual(note);

    input.instance().handleInputChange('note_text', 'big note text');
    input.instance().handleInputChange('customer_visible', true);
    expect(input.state('note')).not.toEqual(note);

    input.instance().onClickReset();
    expect(input.state('note')).toEqual(note);
  });
});
