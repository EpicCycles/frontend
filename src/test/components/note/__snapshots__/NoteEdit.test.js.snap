// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteEdit tests renders the form text correctly with no note 1`] = `
<div>
  <EditModelPage
    model={Object {}}
    modelFields={
      Array [
        Object {
          "displaySize": 400,
          "fieldName": "note_text",
          "header": "Note",
          "type": "textArea",
        },
        Object {
          "fieldName": "customer_visible",
          "header": "Customer View",
          "readOnly": true,
          "type": "checkbox",
        },
        Object {
          "fieldName": "created_date",
          "header": "Date Created",
          "readOnly": true,
          "type": "date_time",
        },
        Object {
          "fieldName": "created_by",
          "header": "Created By",
          "readOnly": true,
          "type": "user",
        },
      ]
    }
    onChange={[Function]}
    persistedModel={Object {}}
  />
  <div
    className="align_right"
  >
    <ModelEditIcons
      componentKey=""
      model={Object {}}
      modelDelete={[MockFunction]}
      modelReset={[Function]}
      modelSave={[MockFunction]}
    />
  </div>
</div>
`;

exports[`NoteEdit tests renders the form text correctly with note 1`] = `
<div>
  <EditModelPage
    model={
      Object {
        "customer_visible": true,
        "id": 23,
        "note_text": "note text here",
      }
    }
    modelFields={
      Array [
        Object {
          "displaySize": 400,
          "fieldName": "note_text",
          "header": "Note",
          "type": "textArea",
        },
        Object {
          "fieldName": "customer_visible",
          "header": "Customer View",
          "readOnly": true,
          "type": "checkbox",
        },
        Object {
          "fieldName": "created_date",
          "header": "Date Created",
          "readOnly": true,
          "type": "date_time",
        },
        Object {
          "fieldName": "created_by",
          "header": "Created By",
          "readOnly": true,
          "type": "user",
        },
      ]
    }
    onChange={[Function]}
    persistedModel={
      Object {
        "customer_visible": true,
        "id": 23,
        "note_text": "note text here",
      }
    }
  />
  <div
    className="align_right"
  >
    <ModelEditIcons
      componentKey={23}
      model={
        Object {
          "customer_visible": true,
          "id": 23,
          "note_text": "note text here",
        }
      }
      modelDelete={[MockFunction]}
      modelReset={[Function]}
      modelSave={[MockFunction]}
    />
  </div>
</div>
`;
